// @ckeditor/ckeditor5-integrations-common/dist/test-utils/index.js
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
const o = /* @__PURE__ */ new Map();
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
const t = /* @__PURE__ */ new Map();
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
const r = "https://cdn.ckeditor.com";
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
function n() {
  s(), i();
}
function s() {
  [...document.querySelectorAll("script")].filter((e) => e.src.startsWith(r)).forEach((e) => {
    t.delete(e.src), e.remove();
  }), delete window.CKEDITOR, delete window.CKEDITOR_PREMIUM_FEATURES, window.CKEDITOR_VERSION = "";
}
function i() {
  [...document.querySelectorAll("link")].filter((e) => e.href.startsWith(r)).forEach((e) => {
    o.delete(e.href), e.remove();
  });
}
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
const c = "https://cdn.ckbox.io";
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
function l() {
  d(), C();
}
function d() {
  [...document.querySelectorAll("script")].filter((e) => e.src.startsWith(c)).forEach((e) => {
    t.delete(e.src), e.remove();
  }), delete window.CKBox;
}
function C() {
  [...document.querySelectorAll("link")].filter((e) => e.href.startsWith(c)).forEach((e) => {
    o.delete(e.href), e.remove();
  });
}
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
function f() {
  n(), l();
}
export {
  f as removeAllCkCdnResources
};
